pkg/cluster/manager.go

```diff
@@ -730,20 +730,24 @@ func (m *Manager) Upgrade(clusterName string, clusterVersion string, opt operato
 			// backup files of the old version
 			tb = tb.BackupComponent(inst.ComponentName(), base.Version, inst.GetHost(), deployDir)
 
-			// copy dependency component if needed
-			switch inst.ComponentName() {
-			case spec.ComponentTiSpark:
-				tb = tb.DeploySpark(inst, version, "" /* default srcPath */, deployDir, m.bindVersion)
-			default:
-				tb = tb.CopyComponent(
-					inst.ComponentName(),
-					inst.OS(),
-					inst.Arch(),
-					version,
-					"", // use default srcPath
-					inst.GetHost(),
-					deployDir,
-				)
+			if deployerInstance, ok := inst.(DeployerInstance); ok {
+				deployerInstance.Deploy(tb, "", deployDir, version, clusterName, clusterVersion)
+			} else {
+				// copy dependency component if needed
+				switch inst.ComponentName() {
+				case spec.ComponentTiSpark:
+					tb = tb.DeploySpark(inst, version, "" /* default srcPath */, deployDir, m.bindVersion)
```

Comment:
It's better to implement the `Deployer` interface for the TiSpark instance.

Reply:
Will do this in later pr